### PR TITLE
[WIP] OpenJFX:  New formula

### PR DESCRIPTION
**Pull request by RandomDSdevel**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?  
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?  
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?  
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?  

-----

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Supersedes Homebrew/homebrew-cask#52957.  Still to be done beyond and/or in the process of filling in the rest of the above checklist:  

* [ ] Addressing some code comments I left in this formula as temporary notes to myself, probably removing almost all of them in the process.  
* [x] Figure out the answer to [this question I've asked on Discourse](https://discourse.brew.sh/t/different-checksum-for-formula-package-download-mirror/3174), particularly within the context of this formula.  
* [ ] Verifying the syntax and intent of this formula's '`install`' method's last statement (the one starting with '`prefix.install …`.'  
* [ ] Linting this formula's `caveats`.  
* [ ] Checking that this formula's test is byte-compiled with and invokes itself with the correct `--module-path`.  

CC @commitay.

### CODE REVIEW

File: Formula/openjfx.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    EOS
+  end
+end
+
+class Openjfx < Formula
+  desc "Open-source, next-generation Java client application platform."
+  homepage "https://openjfx.io/"
+  url "http://hg.openjdk.java.net/openjfx/jfx-dev/rt", :revision => "22c69e868654", :using => :hg
+  sha256 "0be0d342dec0c8ca777ddce3e7ab2afa1e4145181ece4dcaec9f6385cf728fbb"
+  mirror "http://hg.openjdk.java.net/openjfx/jfx-dev/rt/archive/22c69e868654.tar.gz"
```
Comment (vszakats) [inline Formula/openjfx.rb]: Please use HTTPS for `url` and `mirror` as well.

Comment (RandomDSdevel) [reply Formula/openjfx.rb]: @vszakats:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I don't know if HTTPS versions of those downloads are available yet, but I'll check for them.

File: Formula/openjfx.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    <<~EOS
+      This software requires the Xcode CLT in order to build.
+    EOS
+  end
+end
+
+class Openjfx < Formula
+  desc "Open-source, next-generation Java client application platform."
+  homepage "https://openjfx.io/"
+  url "http://hg.openjdk.java.net/openjfx/jfx-dev/rt", :revision => "22c69e868654", :using => :hg
```
Comment (commitay) [inline Formula/openjfx.rb]: Can't use a checkout and tarballs at the same time, mirrors need to have identical content.

https://github.com/javafxports/openjdk-jfx
Did you try the GitHub mirror?

Comment (RandomDSdevel) [reply Formula/openjfx.rb]: @commitay:  

> Can't use a checkout and tarballs at the same time, mirrors need to have identical content.
>
> ⋮

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Thanks, that answers [my Discourse question](https://discourse.brew.sh/t/different-checksum-for-formula-package-download-mirror/3174).  Could you also leave this information there, please, or let me quote your response here there?  In any case, which of the various download sources would you prefer?  

> ⋮
>
> https://github.com/javafxports/openjdk-jfx
> Did you try the GitHub mirror?

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Yes, that's the second download mirror URL.

**Comment by fxcoudert** [comment]
Thanks @RandomDSdevel for the pull request. We really prefer that formulas build locally and pass audit before they are submitted as pull requests, though, as we tend to have too many open PRs at early stages of progress, which make triaging and reviewing hard (most of the review comments we leave are really fixed by reading audit messages and fixing them).

**Comment by RandomDSdevel** [comment]
@fxcoudert:  

> Thanks @RandomDSdevel for the pull request. …

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;You're welcome.  

> …We really prefer that formulas build locally and pass audit before they are submitted as pull requests, though, as we tend to have too many open PRs at early stages of progress, which make triaging and reviewing hard (most of the review comments we leave are really fixed by reading audit messages and fixing them).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;That's understandable, and I would personally have preferred if I could have made that happen for this formula as well.  I had some concerns with this formula that I wasn't entirely sure I could address on my own, however, despite otherwise having completed putting it together, so I went ahead and submitted the work I had for interim staging.  (If one could associate an issue or pull-request–like discussion thread with a branch in their downstream fork of a repository before pushing it upstream as a PR, then this would likely have been a situation that called for doing just that.  As far as I know, though, GitHub doesn't seem to have anything approaching that sort of functionality.)  I did this with the assumption that I would be permitted/required to perform a build and final audit locally once all open items on the to-do checklist I put in this PR's OP are taken care of and before my commits could be merged.

File: Formula/openjfx.rb (hunk 2 above)
Comment (RandomDSdevel) [reply Formula/openjfx.rb]: @commitay:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorry if it wasn't clear, but that was only with respect to that Discourse thread.  I still need to know:  

> …which of the various download sources would you prefer?

Comment (commitay) [reply Formula/openjfx.rb]: Depends really, how is it versioned? The current revision doesn't look like it corresponds to a tag?

Comment (RandomDSdevel) [reply Formula/openjfx.rb]: &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Well, I was thinking that the version number would need to match the '11+28' of the current JDK release, but I checked the build distributed on the OpenJFX as part of its official standalone JavaFX SDK package…and, according to that ZIP artifact's contained '`lib/javafx.properties`' file, said build's version is '11+26.'  The GitHub repository mirror has a tag for that, so I'll just use one of the archives associated with it.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I'm running a test build locally myself now.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I've now run a successful build locally.  Looks like I got the test module path wrong somehow, though…

**Comment by RandomDSdevel** [comment]
@commitay, @fxcoudert:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I'm getting closer!  Here's a status update:  

- `brew style` and `brew audit --strict` both pass with flying colors now.  
- Remaining things to do:  
  - There are still a couple comments that I've left behind, but I think that they're less personal and more documentation.  I can touch up and/or remove these as needed.  
  - The `caveats` I've supplied might possibly read better if reworded, but I'm not exactly sure what the ideal ones would look like.  
  - I still haven't gotten this formula's test to properly link and pass; it currently fails to find OpenJFX even after `brew` installs it.  I could well still be overlooking something related to formula package installation and/or test byte-compilation and/or execution.  

What are your recommendations?

**Comment by RandomDSdevel** [comment]
@commitay, @fxcoudert:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Hmm, it looks like installing the contents of'`build_path/"modular_sdk"`' as instructed by the OpenJFX build guide grabs a bit more than is included in upstream's pre-built SDK ZIP.  That explains why the '`--module-path`' against which my test here is byte-compiled and linked isn't correct.  Unfortunately, I am leaving with my family for a weekend vacation in an hour and a half or so and thus won't be able to work on this until after we get back late on Monday.  I'll sort through these remaining test issues then.

**Comment by RandomDSdevel** [comment]
@commitay, @fxcoudert:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorry to have been keeping this PR in limbo for so long after I got back; I started poking at another of my outstanding Homebrew issues/PRs-under-construction.  I'll get back to this as soon as I can.  In the meantime, please give me your thoughts on the to-do list of items that remain to be addressed from [that earlier comment of mine](https://github.com/Homebrew/homebrew-core/pull/32864#issuecomment-429178333).

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Oops, I nearly forgot to mention explicitly that I've started working on trying to get everything installed correctly again.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A-_ha!_  I found most of what I was looking for all along inside the build results:  everything gets packaged up into '`buildpath/"artifacts/javafx-sdk-#{version}/"`',  so the formula can just install that folder's contents instead of fishing all over the build path for everything that lands there.  How I didn't figure this out sooner, I don't rightly know; it's so _obvious_ now…😅.   Anyway, not _quite_ everything offered by upstream's pre-packaged SDK archive is in there, but I'm confident that fixing that is just a matter of tweaking some build options.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Huh, [looks like I got tripped up by some misleadingly named build properties](http://mail.openjdk.java.net/pipermail/openjfx-dev/2018-November/022872.html); those should do the trick.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I should mention that getting OpenJFX to build properly _should_ fix the test failure now that I'm installing the right stuff.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I've pushed what I have, but don't expect it to pass CI.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(_Bonks stale-bot on the nose like he would a Star Wars pit droid._)  There's still a build error/issue I have yet to mitigate, you silly bucket of bolts.

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This PR now depends on #35046.

**Comment by equal-l2** [comment]
Sorry for jumping in but can you use the bundled gradlew,  instead of adding a new dependency which will be used only by this?
(I'm not sure Homebrew allows gradlew, which requires an additional downloading, though)

**Comment by SMillerDev** [comment]
@equal-l2 don't be sorry for trying to make homebrew better. If we can use the versions of gradle we ship already we'd prefer those over things that automagically download themselves.

**Comment by RandomDSdevel** [comment]
@equal-l2:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;As soon as OpenJFX can build successfully using Gradle v5.x, it can be switched to using Homebrew's pre-existing formula for that instead and I can submit a PR to counter #35046's result.  In the meantime, the project still requires a Gradle version greater than or equal to 4.8.something and less than 5.0, so, as @SMillerDev said, Homebrew's going to have to ship a Gradle package with a version satisfying those constraints for now.  (I'm _actually_ seeing warnings with v4.10.3 as it is _already_, as OpenJFX _really_ only strictly wants Gradle v4.8.x, but I was counting on that possibly just being overly cautious of upstream…)  

---

@SMillerDev:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On another note, I've searched around for clues in some other, older Homebrew issues and seen that problems like the one I'm seeing [above](https://github.com/Homebrew/homebrew-core/pull/32864#issuecomment-437618905) arise when build systems get confused and pick things up from both Xcode's internal SDK and the Command-Line Tools package's global one instead of choosing one or the other.  Might you have any advice on how to resolve that?

**Comment by RandomDSdevel** [comment]
@fxcoudert:  

> OpenJFX is actively maintained and has many more recent releases, …

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Yes, but OpenJFX v11.0.1 is what this PR adds to Homebrew Core and is currently the latest release.  OpenJFX v12 is available only in the form of early-access (beta) builds at this time.  

> …hopefully they will support the latest gradle soon (or already do).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I found the following relevant issues open upstream:  

- [JDK-8210509:  "maven publishing task causes gradle deprecation warning"](https://bugs.openjdk.java.net/browse/JDK-8210509)
- [JDK-8215686:  "FX build fails using gradle 5"](https://bugs.openjdk.java.net/browse/JDK-8215686) (As noted in that issue's initial description, a fix for this isn't targeted for inclusion in OpenJFX until v13.)  (See also javafxports/openjdk-jfx#313.)  

as well as an additional old, closed issue:  

- [JDK-8199071:  "Fix gradle deprecation warnings"](https://bugs.openjdk.java.net/browse/JDK-8199071)

None of these sources suggest that the OpenJFX community is in much of any hurry to move away from Gradle v4.8.x and on to Gradle v5.x in the near future, which is why I was having this PR depend on #35046.  On another note, none of this addresses the SDK-path confusion–related build error documented [earlier](https://github.com/Homebrew/homebrew-core/pull/32864#issuecomment-437618905) that I continue to see, unfortunately.

**Pull request closed without merging**